File: morning_exercises/day19.cs
class BinaryNode4(int val = 0, BinaryNode4? left = null, BinaryNode4? right = null)
{
    public int val = val;
    public BinaryNode4? left = left;
    public BinaryNode4? right = right;

    public static int MaxPathSumHelper(BinaryNode4? root, List<int> globalMax)
    {
        if (root==null) return 0;

        int leftBest = Math.Max(0, MaxPathSumHelper(root.left, globalMax));
        int rightBest = Math.Max(0, MaxPathSumHelper(root.right, globalMax));

        globalMax[0] = Math.Max(globalMax[0], root.val+leftBest+rightBest);

        return root.val + Math.Max(leftBest, rightBest);
    }

    public static int MaxPathSum(BinaryNode4? root)
    {
        List<int> globalMax = [int.MinValue];

        MaxPathSumHelper(root, globalMax);

        return globalMax[0];
    }
}

File: morning_exercises/day19.ts
class BinaryNode4 {
    constructor(
        public val: number = 0,
        public left: BinaryNode4 | null = null,
        public right: BinaryNode4 | null = null,
    ) {}

    static maxPathSumHelper(root: BinaryNode4 | null, globalMax: number[]): number {
        if (root==null) return 0;

        let leftBest: number = Math.max(0, BinaryNode4.maxPathSumHelper(root.left, globalMax));
        let rightBest: number = Math.max(0, BinaryNode4.maxPathSumHelper(root.right, globalMax));

        globalMax[0] = Math.max(globalMax[0], root.val+leftBest+rightBest);
        return root.val+Math.max(leftBest, rightBest);
    }

    static maxPathSum(root: BinaryNode4 | null): number {
        let globalMax: number[] = [-Infinity];

        BinaryNode4.maxPathSumHelper(root, globalMax);

        return globalMax[0];
    }
}

{
    const root = new BinaryNode4(11);
    root.left = new BinaryNode4(2);
    root.right = new BinaryNode4(3);
    root.left.left = new BinaryNode4(4);
    root.left.right = new BinaryNode4(5);

    console.log(BinaryNode4.maxPathSum(root));
}